engines/cufile.c

```diff
@@ -0,0 +1,586 @@
+/*
+ * Copyright (c)2020 System Fabric Works, Inc. All Rights Reserved.
+ * mailto:info@systemfabricworks.com
+ *
+ * License: GPLv2, see COPYING.
+ *
+ * cufile engine
+ *
+ * fio I/O engine using the NVIDIA cuFile API.
+ *
+ */
+
+#include <stdlib.h>
+#include <unistd.h>
+#include <errno.h>
+#include <string.h>
+#include <sys/time.h>
+#include <sys/resource.h>
+#include <cufile.h>
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <pthread.h>
+
+#include "../fio.h"
+#include "../lib/pow2.h"
+#include "../optgroup.h"
+#include "../lib/memalign.h"
+
+#define ALIGNED_4KB(v) (((v) & 0x0fff) == 0)
+
+#define LOGGED_BUFLEN_NOT_ALIGNED     0x01
+#define LOGGED_OFFSET_NOT_ALIGNED     0x02
+#define LOGGED_GPU_OFFSET_NOT_ALIGNED 0x03
+#define GPU_ID_SEP ":"
+
+struct cufile_options {
+	struct thread_data *td;
+	char               *gpu_ids;       /* colon-separated list of GPU ids,
+										  one per job */
+	void               *cu_mem_ptr;    /* GPU memory */
+	void               *junk_buf;      /* buffer to simulate cudaMemcpy with
+										  posix I/O write */
+	int                 my_gpu_id;     /* GPU id to use for this job */
+	unsigned int        use_posixio;   /* Use POSIX I/O instead of cuFile API */
+	size_t              total_mem;     /* size for cu_mem_ptr and junk_buf */
+	int                 logged;        /* bitmask of log messages that have
+										  been output, prevent flood */
+};
+
+struct fio_cufile_data {
+	CUfileDescr_t  cf_descr;
+	CUfileHandle_t cf_handle;
+};
+
+struct fio_option options[] = {
+	{
+		.name	= "cufile_gpu_ids",
+		.lname	= "cufile gpu ids",
+		.type	= FIO_OPT_STR_STORE,
+		.off1	= offsetof(struct cufile_options, gpu_ids),
+		.help	= "GPU IDs, one per job, separated by " GPU_ID_SEP,
+		.category = FIO_OPT_C_ENGINE,
+		.group	= FIO_OPT_G_CUFILE,
+	},
+	{
+		.name	= "cufile_posixio",
+		.lname	= "cufile posixio",
+		.type	= FIO_OPT_BOOL,
+		.off1	= offsetof(struct cufile_options, use_posixio),
+		.help	= "Use POSIX I/O instead of cuFile API",
+		.def    = "0",
+		.category = FIO_OPT_C_ENGINE,
+		.group	= FIO_OPT_G_CUFILE,
+	},
+	{
+		.name	= NULL,
+	},
+};
+
+static int running = 0;
+static int cufile_initialized = 0;
+static pthread_mutex_t running_lock = PTHREAD_MUTEX_INITIALIZER;
+
+#define check_cudaruntimecall(fn, rc)								\
+	do {															\
+		cudaError_t res = fn;										\
+		if (res != cudaSuccess) {									\
+			const char *str = cudaGetErrorName(res);				\
+			log_err("cuda runtime api call failed %s:%d : %s\n",	\
+					#fn, __LINE__, str);							\
+			rc = -1;												\
+		} else {													\
+			rc = 0;													\
+		}															\
+	} while(0)
```

Comment:
I'll assume GitHub has done a poor job of showing how these line up :-)

Reply:
github is using a different tab width than my editor. I'll change the whitespace to the line continuation backslash to use spaces. That should tame it.